### PR TITLE
add functions to create carbon file from array_it

### PR DESCRIPTION
**Pull request by Jasper-Ben**
Adds functions for creating a carbon file from an array iterator.
Cherry-picked commit https://github.com/jaksonlabs/jakson/pull/4/commits/442493207ded119bef22e21bb1e850c9e5b640b7, so you might want to review https://github.com/jaksonlabs/jakson/pull/4 first.
Also cherry-picked commit 656b21689f7295f3726bb9f81571f8a300ca1661 from https://github.com/jaksonlabs/jakson/pull/1.

### CODE REVIEW

File: src/jakson/carbon.c [hunk 1]
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
...
+
+    bool isnull;
+    carbon_field_type_e type;
+
+    carbon_object_it_rewind(it_in);
+
+    while (carbon_object_it_next(it_in)) {
+        u64 keylen;
+        const char *_key = carbon_object_it_prop_name(&keylen, it_in);
+        const char *key = strndup(_key, keylen);
```
Comment (Jasper-Ben) [inline src/jakson/carbon.c]: memleak, fixme

File: src/jakson/carbon.c [hunk 2]
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
...
+        }
+        else if (carbon_field_type_is_floating_atom(type)) {
+            float val;
+            carbon_object_it_float_value(&isnull, &val, it_in);
+            carbon_insert_prop_float(ins, key, val);
+        }
+        else if (carbon_field_type_is_string(type)) {
+            u64 strlen;
+            const char *_str = carbon_object_it_string_value(&strlen, it_in);
+            const char *str = strndup(_str, strlen);
```
Comment (Jasper-Ben) [inline src/jakson/carbon.c:385]: memleak, fixme

File: src/jakson/carbon.c [hunk 3]
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
...
+            carbon_insert_prop_string(ins, key, str);
+        }
+        else if (carbon_field_type_is_binary(type)) {
+            carbon_binary bin;
+            carbon_object_it_binary_value(&bin, it_in);
+            const char *mime = strndup(bin.mime_type, bin.mime_type_strlen);
```
Comment (Jasper-Ben) [inline src/jakson/carbon.c:391]: memleak, fixme

File: src/jakson/carbon.c [hunk 4]
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
...
+        }
+        else if (carbon_field_type_is_floating_atom(type)) {
+            float val;
+            carbon_array_it_float_value(&isnull, &val, it_in);
+            carbon_insert_float(ins, val);
+        }
+        else if (carbon_field_type_is_string(type)) {
+            u64 strlen;
+            const char *_str = carbon_array_it_string_value(&strlen, it_in);
+            const char *str = strndup(_str, strlen);
```
Comment (Jasper-Ben) [inline src/jakson/carbon.c]: memleak, fixme

File: src/jakson/carbon.c [hunk 5]
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
...
+            carbon_insert_string(ins, str);
+        }
+        else if (carbon_field_type_is_binary(type)) {
+            carbon_binary bin;
+            carbon_array_it_binary_value(&bin, it_in);
+            const char *mime = strndup(bin.mime_type, bin.mime_type_strlen);
```
Comment (Jasper-Ben) [inline src/jakson/carbon.c]: memleak, fixme

File: src/jakson/carbon.c [hunk 6]
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
...
+
+            const void *vals = carbon_column_it_values(&type, &nvalues, nested_column);
+            if (carbon_field_type_is_signed(type)) {
+                nested_ins = carbon_insert_column_begin(&state, ins, CARBON_COLUMN_TYPE_I64, 1024);
+                for (u64 i = 0; i < nvalues; i++) {
+                    if (carbon_column_it_value_is_null(nested_column, i)) {
+                        carbon_insert_null(nested_ins);
+                    }
+                    else {
+                        carbon_insert_i64(nested_ins, *((i64*)&(vals[i])));
```
Comment (Jasper-Ben) [inline src/jakson/carbon.c:303]: this will break, as it will try to open i8 mem as i64 etc.

**Comment by heregoesmarcel** [comment]
Currently, your added `carbon_from_array_it()` seems to initialize a new carbon document with a fixed key and options. This should to my knowledge be parameterized.

File: src/jakson/carbon.c [hunk 7]
```diff
@@ -188,6 +188,304 @@ bool carbon_from_raw_data(carbon *doc, err *err, const void *data, u64 len)
...
+        }
+        else if (carbon_field_type_is_null(type)) {
+            carbon_insert_null(ins);
+        }
+        else if (carbon_field_type_is_array_or_subtype(type)) {
+            carbon_array_it *nested_array = carbon_array_it_array_value(it_in);
+            carbon_insert *nested_ins;
+            carbon_insert_array_state state;
+
+            nested_ins = carbon_insert_array_begin(&state, ins, 1024);
```
Comment (Jasper-Ben) [inline src/jakson/carbon.c]: use array_len param, instead of fixed size

**Pull request closed without merging**